File: UCA-4/Linked_List/Java/28_1.java
//Implementation of Doubly Linked Lists
import java.util.*;

class Node{
	int data;
	Node next;
	Node prev;

	public Node(int data,Node next,Node prev){
		this.data = data;
		this.next = next;
		this.prev = prev;
	}
}

class LL{
	public Node head = null;
	public void insertNode(int value){
		if(head == null){
			head = new Node(value,null,null);
		}else{
			Node current = head;
			while(current.next!=null){
				current = current.next;
			}
			current.next = new Node (value,null,current);
		}
	}

	public void traverse(){
		Node current = head;
		while(current!=null){
			System.out.println(current.data);
			current = current.next;
		}
	}

	public void insertNodeBeg(int value){
		if (head==null) {
			head = new Node(value,null,null);
		}else{
			Node temp = new Node(value,head,null);
			head = temp;
		}
	}

	public int countNodes(){
		int count = 0;
		Node current = head;
		while(current!=null){
			count++;
			current = current.next;
		}
		return count;
	}

}

class Test{
	public static void main(String[] args) {
		LL obj1 = new LL();
		obj1.insertNode(5);
		obj1.insertNode(6);
		obj1.insertNodeBeg(7);
		obj1.traverse();
		System.out.println(obj1.countNodes());
	}
}

File: UCA-4/Linked_List/C/28_1.c
//Implement A Linked List and Calculate its Size
// 28 Sept,2018
#include <stdio.h>
#include <stdlib.h>

struct Node{
	int data;
	struct Node *next;
};

struct Node* createNode(int value){
	struct Node *temp = (struct Node*) malloc(sizeof(struct Node));
	temp->data = value;
	temp->next = NULL;
	return temp;
};

struct Node *head = NULL;

void insertNode(int value){
	if(head==NULL){
		head = createNode(value);
	}else{
		struct Node* current = head;
        while(current->next!=NULL){
            current = current->next;
        }
		current->next=createNode(value);
	}
}

void insertNodeBeg(int value){
    if(head==NULL){
        head = createNode(value);
    }else{
        struct Node *temp = createNode(value);
        temp->next=head;
        head = temp;
    }
}

void traverse(){
	struct Node* current = head;
	while(current!=NULL){
		printf("%d\n",current->data);
		current=current->next;
	}
    free(current);
}

int calculateSize(){
	int count = 0;
	struct Node* current = head;
	while(current!=NULL){
		count++;
		current=current->next;
	}
	return count;
    free(current);
}

int main(int argc, char const *argv[])
{
	insertNode(5);
	insertNode(6);
    insertNodeBeg(7);
    traverse();
	printf("%d\n",calculateSize());
	return 0;
}
